### PR TITLE
Add support for directional registers to enable lossless Cirq-FT interop

### PR DESCRIPTION
**Pull request by tanujkhattar**
This is (hopefully) the final PR in a series of PRs to enable "lossless" Cirq-FT interop. This PR adds support for directional registers in Cirq-FT interop, which was added in Cirq as part of https://github.com/quantumlib/Cirq/pull/6300

Using the improvements made in this PR, a user should ideally be able to seamlessly import non-trivial FT gadgets written in either Cirq-FT as `GateWithRegister`s or in qualtran as `Bloq`s  in one or the other framework. 

Two non-trivial examples are shown in the `cirq_interop.ipynb` notebook:
1. `Bloq` -> `Cirq-FT` -> `Bloq` roundtrip using `ModExp` bloq.
2. `Cirq-FT` -> `Bloq` -> `Cirq-FT` roundtrip using `QROM` gate. 

Two big TODOs are to add more tests and more documentation. Another potential TODO is to see if we can simplify the `_cirq_interop` code structure by splitting into two files - one for `_cirq_to_bloqs.py` and another for `_bloqs_to_cirq.py`. 

For now, I'm happy to see this working. Let's do a first round of reviews and I can then fix some of the TODOs either as part of this PR or as part of a follow-up. 

cc @fdmalone @mpharrigan

### CODE REVIEW

File: qualtran/cirq_interop/_cirq_interop.py [hunk 1]
```diff
@@ -426,38 +417,85 @@ def _cbloq_to_cirq_circuit(
...
+        for reg in signature.lefts()
+        for idx in reg.all_idxs()
+    }
     moments: List[cirq.Moment] = []
     for binsts in nx.topological_generations(binst_graph):
         moment: List[cirq.Operation] = []
 
         for binst in binsts:
-            if isinstance(binst, DanglingT):
+            if binst == LeftDangle:
```
Comment (mpharrigan) [inline qualtran/cirq_interop/_cirq_interop.py:431]: elsewhere I've been using `binst is LeftDangle`

**Comment by mpharrigan** [comment]
is it possible to write unit tests for subsets of the functionalities (i.e. some of the helper functions)

I think splitting into multiple files is called for, yes.

File: qualtran/cirq_interop/_cirq_interop.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/_cirq_interop.py]: changed.

File: qualtran/cirq_interop/_bloq_to_cirq.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    ) -> cirq.OP_TREE:
+        """Implementation of the GatesWithRegisters decompose method.
+
+        This delegates to `self.bloq.decompose_bloq()` and converts the result to a cirq circuit.
+
+        Args:
+            context: `cirq.DecompositionContext` stores options for decomposing gates (eg:
+                cirq.QubitManager).
+            **quregs: Sequences of cirq qubits as expected for the legacy register shims
+            of the bloq's registers.
```
Comment (mpharrigan) [inline qualtran/cirq_interop/_bloq_to_cirq.py:103]: update docstring?

**Review by mpharrigan — APPROVED** [review]
I did some git trickery locally to try to get a reasonable looking diff lol

But LGTM! Complexity seems well isolated into functions

File: qualtran/cirq_interop/_bloq_to_cirq.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/_bloq_to_cirq.py]: done

**Comment by tanujkhattar** [comment]
Added more tests and also fixed a bug (aka extended functionality for directional registers) in `CirqGateAsBloq.add_my_tensors` method (see the added test `test_cirq_gate_as_bloq_tensor_contract_for_and_gate`).

Merging now!

**Pull request merged**